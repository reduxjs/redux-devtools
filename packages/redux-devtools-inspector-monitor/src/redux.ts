import { Action } from 'redux';
import { DevtoolsInspectorProps } from './DevtoolsInspector';

const UPDATE_MONITOR_STATE =
  '@@redux-devtools-inspector-monitor/UPDATE_MONITOR_STATE';

export interface UpdateMonitorStateAction {
  type: typeof UPDATE_MONITOR_STATE;
  monitorState: Partial<DevtoolsInspectorState>;
}
export function updateMonitorState(
  monitorState: Partial<DevtoolsInspectorState>,
): UpdateMonitorStateAction {
  return { type: UPDATE_MONITOR_STATE, monitorState };
}

export type DevtoolsInspectorAction = UpdateMonitorStateAction;

export interface DevtoolsInspectorState {
  selectedActionId: number | null;
  startActionId: number | null;
  inspectedActionPath: (string | number)[];
  inspectedStatePath: (string | number)[];
  tabName: string;
  searchValue?: string;
}

export const DEFAULT_STATE: DevtoolsInspectorState = {
  selectedActionId: null,
  startActionId: null,
  inspectedActionPath: [],
  inspectedStatePath: [],
  tabName: 'Diff',
};

function reduceUpdateState(
  state: DevtoolsInspectorState,
  action: DevtoolsInspectorAction,
) {
  return action.type === UPDATE_MONITOR_STATE
    ? {
        ...state,
        ...action.monitorState,
      }
    : state;
}

export function reducer<S, A extends Action<unknown>>(
  props: DevtoolsInspectorProps<S, A>,
  state = DEFAULT_STATE,
  action: DevtoolsInspectorAction,
) {
  return {
    ...reduceUpdateState(state, action),
  };
}
