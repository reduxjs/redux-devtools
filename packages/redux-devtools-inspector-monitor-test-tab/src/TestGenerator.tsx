import React, { PureComponent, Component, ReactNode } from 'react';
import { stringify } from 'javascript-stringify';
import objectPath from 'object-path';
import jsan from 'jsan';
import diff, { Event } from 'simple-diff';
import es6template from 'es6template';
import { Editor } from '@redux-devtools/ui';
import { TabComponentProps } from '@redux-devtools/inspector-monitor';
import { Action } from 'redux';
import { AssertionLocals, DispatcherLocals, WrapLocals } from './types';

export const fromPath = (path: (string | number)[]) =>
  path.map((a) => (typeof a === 'string' ? `.${a}` : `[${a}]`)).join('');

function getState<S>(
  s: { state: S; error?: string } | undefined,
  // eslint-disable-next-line @typescript-eslint/ban-types
  defaultValue?: {}
) {
  if (!s) return defaultValue;
  return JSON.parse(jsan.stringify(s.state));
}

export function compare<S>(
  s1: { state: S; error?: string } | undefined,
  s2: { state: S; error?: string },
  cb: (value: { path: string; curState: number | string | undefined }) => void,
  // eslint-disable-next-line @typescript-eslint/ban-types
  defaultValue?: {}
) {
  const paths: string[] = []; // Already processed
  function generate(
    event: Event | { type: 'move-item'; newPath: (string | number)[] }
  ) {
    let curState: number | string | undefined;
    let path = fromPath(event.newPath);

    if (event.type === 'remove-item' || event.type === 'move-item') {
      if (paths.length && paths.indexOf(path) !== -1) return;
      paths.push(path);
      // eslint-disable-next-line @typescript-eslint/ban-types
      const v = objectPath.get(s2.state as unknown as object, event.newPath);
      curState = v.length;
      path += '.length';
    } else if (event.type === 'add-item') {
      generate({ type: 'move-item', newPath: event.newPath });
      path += `[${event.newIndex}]`;
      curState = stringify(event.newValue);
    } else {
      curState = stringify(event.newValue);
    }

    // console.log(`expect(store${path}).toEqual(${curState});`);
    cb({ path, curState });
  }

  diff(
    getState(s1, defaultValue),
    getState(s2, defaultValue) /* , { idProp: '*' } */
  ).forEach(generate);
}

interface Props<S, A extends Action<unknown>>
  extends Omit<TabComponentProps<S, A>, 'monitorState' | 'updateMonitorState'> {
  name?: string;
  isVanilla?: boolean;
  wrap?: string | ((locals: WrapLocals) => string);
  dispatcher?: string | ((locals: DispatcherLocals) => string);
  assertion?: string | ((locals: AssertionLocals) => string);
  useCodemirror: boolean;
  indentation?: number;
  header?: ReactNode;
}

export default class TestGenerator<
  S,
  A extends Action<unknown>
> extends (PureComponent || Component)<Props<S, A>> {
  getMethod(action: A) {
    let type: string = action.type as string;
    if (type[0] === '┗') type = type.substr(1).trim();
    const args = (action as unknown as { arguments: unknown[] }).arguments
      ? (action as unknown as { arguments: unknown[] }).arguments
          .map((arg) => stringify(arg))
          .join(',')
      : '';
    return `${type}(${args})`;
  }

  getAction(action: A) {
    if (action.type === '@@INIT') return '{}';
    return stringify(action);
  }

  generateTest() {
    const {
      computedStates,
      actions,
      selectedActionId,
      startActionId,
      isVanilla,
      name,
    } = this.props;

    if (!actions || !computedStates || computedStates.length < 1) return '';

    let { wrap, assertion, dispatcher, indentation } = this.props;
    if (typeof assertion === 'string')
      assertion = es6template.compile(assertion);
    if (typeof wrap === 'string') {
      const ident = /\n.+\$\{assertions}/.exec(wrap);
      if (ident) indentation = ident[0].length - 13;
      wrap = es6template.compile(wrap);
    }
    if (typeof dispatcher === 'string')
      dispatcher = es6template.compile(dispatcher);

    let space = '';
    if (indentation) space = Array(indentation).join(' ');

    let r = '';
    let isFirst = true;
    let i;
    if (startActionId !== null) i = startActionId;
    else if (selectedActionId !== null) i = selectedActionId;
    else i = computedStates.length - 1;
    const startIdx = i > 0 ? i : 1;

    const addAssertions = ({
      path,
      curState,
    }: {
      path: string;
      curState: number | string | undefined;
    }) => {
      r += `${space}${(assertion as (locals: AssertionLocals) => string)({
        path,
        curState,
      })}\n`;
    };

    while (actions[i]) {
      if (
        !isVanilla ||
        /* eslint-disable-next-line no-useless-escape */
        /^┗?\s?[a-zA-Z0-9_@.\[\]-]+?$/.test(actions[i].action.type as string)
      ) {
        if (isFirst) isFirst = false;
        else r += space;
        if (!isVanilla || (actions[i].action.type as string)[0] !== '@') {
          r +=
            (dispatcher as (locals: DispatcherLocals) => string)({
              action: !isVanilla
                ? this.getAction(actions[i].action)
                : this.getMethod(actions[i].action),
              prevState:
                i > 0 ? stringify(computedStates[i - 1].state) : undefined,
            }) + '\n';
        }
        if (!isVanilla) {
          addAssertions({
            path: '',
            curState: stringify(computedStates[i].state),
          });
        } else {
          compare(
            computedStates[i - 1],
            computedStates[i],
            addAssertions,
            isVanilla && {}
          );
        }
      }
      i++;
      if (i > selectedActionId!) break;
    }

    r = r.trim();
    if (wrap) {
      if (!isVanilla) r = wrap({ name, assertions: r });
      else {
        r = wrap({
          name: /^[a-zA-Z0-9_-]+?$/.test(name as string) ? name : 'Store',
          actionName:
            (selectedActionId === null || selectedActionId > 0) &&
            actions[startIdx]
              ? (actions[startIdx].action.type as string).replace(
                  /[^a-zA-Z0-9_-]+/,
                  ''
                )
              : 'should return the initial state',
          initialState: stringify(computedStates[startIdx - 1].state),
          assertions: r,
        });
      }
    }
    return r;
  }

  render() {
    const code = this.generateTest();

    if (!this.props.useCodemirror) {
      return (
        <textarea
          style={{ padding: '10px', width: '100%', height: '100%' }}
          defaultValue={code}
        />
      );
    }

    return <Editor value={code} />;
  }

  static defaultProps = {
    useCodemirror: true,
    selectedActionId: null,
    startActionId: null,
  };
}
