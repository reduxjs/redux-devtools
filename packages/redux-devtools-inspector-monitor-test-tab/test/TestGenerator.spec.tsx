import React from 'react';
import { render, screen } from '@testing-library/react';
import { PerformAction } from '@redux-devtools/core';
import { Action } from 'redux';
import TestGenerator from '../src/TestGenerator';
import fnTemplate from '../src/redux/mocha';
import strTemplate from '../src/redux/mocha/template';
import fnVanillaTemplate from '../src/vanilla/mocha';
import strVanillaTemplate from '../src/vanilla/mocha/template';

const actions: { [actionId: number]: PerformAction<Action<unknown>> } = {
  0: {
    type: 'PERFORM_ACTION',
    action: { type: '@@INIT' },
    timestamp: 0,
    stack: undefined,
  },
  1: {
    type: 'PERFORM_ACTION',
    action: { type: 'INCREMENT_COUNTER' },
    timestamp: 0,
    stack: undefined,
  },
};

const computedStates = [{ state: { counter: 0 } }, { state: { counter: 1 } }];

const TestGeneratorAsAny = TestGenerator as any;

describe('TestGenerator component', () => {
  it('should show warning message when no params provided', () => {
    render(<TestGeneratorAsAny useCodemirror={false} />);
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });

  it('should be empty when no actions provided', () => {
    render(
      <TestGeneratorAsAny
        assertion={fnTemplate.assertion}
        dispatcher={fnTemplate.dispatcher}
        wrap={fnTemplate.wrap}
        useCodemirror={false}
      />
    );
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });

  it("should match function template's test for first action", () => {
    render(
      <TestGeneratorAsAny
        assertion={fnTemplate.assertion}
        dispatcher={fnTemplate.dispatcher}
        wrap={fnTemplate.wrap}
        actions={actions}
        computedStates={computedStates}
        selectedActionId={1}
        useCodemirror={false}
      />
    );
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });

  it("should match string template's test for first action", () => {
    render(
      <TestGeneratorAsAny
        assertion={strTemplate.assertion}
        dispatcher={strTemplate.dispatcher}
        wrap={strTemplate.wrap}
        useCodemirror={false}
        actions={actions}
        computedStates={computedStates}
        selectedActionId={1}
      />
    );
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });

  it('should generate test for the last action when selectedActionId not specified', () => {
    render(
      <TestGeneratorAsAny
        assertion={fnTemplate.assertion}
        dispatcher={fnTemplate.dispatcher}
        wrap={fnTemplate.wrap}
        actions={actions}
        computedStates={computedStates}
        useCodemirror={false}
      />
    );
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });

  it('should generate test for vanilla js class', () => {
    render(
      <TestGeneratorAsAny
        assertion={fnVanillaTemplate.assertion}
        dispatcher={fnVanillaTemplate.dispatcher}
        wrap={fnVanillaTemplate.wrap}
        actions={actions}
        computedStates={computedStates}
        selectedActionId={1}
        isVanilla
        name="SomeStore"
        useCodemirror={false}
      />
    );
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });

  it('should generate test for vanilla js class with string template', () => {
    render(
      <TestGeneratorAsAny
        assertion={strVanillaTemplate.assertion}
        dispatcher={strVanillaTemplate.dispatcher}
        wrap={strVanillaTemplate.wrap}
        actions={actions}
        computedStates={computedStates}
        selectedActionId={1}
        isVanilla
        name="SomeStore"
        useCodemirror={false}
      />
    );
    expect(screen.getByRole('textbox')).toMatchSnapshot();
  });
});
