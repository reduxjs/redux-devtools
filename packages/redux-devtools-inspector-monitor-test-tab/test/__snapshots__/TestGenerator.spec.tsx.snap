// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestGenerator component should be empty when no actions provided 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
/>
`;

exports[`TestGenerator component should generate test for the last action when selectedActionId not specified 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
>
  import expect from 'expect';
import reducers from '../../reducers';

describe('reducers', () =&gt; {
  it('should handle actions', () =&gt; {
    let state;
    state = reducers({counter:0}, {type:'INCREMENT_COUNTER'});
expect(state).toEqual({counter:1});
  });
});

</textarea>
`;

exports[`TestGenerator component should generate test for vanilla js class 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
>
  import expect from 'expect';
import SomeStore from '../../stores/SomeStore';

describe('SomeStore', () =&gt; {
  it('INCREMENT_COUNTER', () =&gt; {
    const store = new SomeStore({counter:0});
    INCREMENT_COUNTER();
expect(store.counter).toEqual(1);
  });
});

</textarea>
`;

exports[`TestGenerator component should generate test for vanilla js class with string template 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
>
  import expect from 'expect';
import SomeStore from '../../stores/SomeStore';

describe('SomeStore', () =&gt; {
  it('INCREMENT_COUNTER', () =&gt; {
    const store = new SomeStore({counter:0});
    INCREMENT_COUNTER();
    expect(store.counter).toEqual(1);
  });
});

</textarea>
`;

exports[`TestGenerator component should match function template's test for first action 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
>
  import expect from 'expect';
import reducers from '../../reducers';

describe('reducers', () =&gt; {
  it('should handle actions', () =&gt; {
    let state;
    state = reducers({counter:0}, {type:'INCREMENT_COUNTER'});
expect(state).toEqual({counter:1});
  });
});

</textarea>
`;

exports[`TestGenerator component should match string template's test for first action 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
>
  import expect from 'expect';
import reducers from '../../reducers';

describe('reducers', () =&gt; {
  it('should handle actions', () =&gt; {
    let state;
    state = reducers({counter:0}, {type:'INCREMENT_COUNTER'});
    expect(state).toEqual({counter:1});
  });
});

</textarea>
`;

exports[`TestGenerator component should show warning message when no params provided 1`] = `
<textarea
  style="padding: 10px; width: 100%; height: 100%;"
/>
`;
